### PR TITLE
Csrf fix [WIP]

### PR DESCRIPTION
**Pull request by ieaves**
Changes:
* Substantial improvements to server configuration supporting much richer user customization over the running environment
* Fixes CORS for applications deployed outside of localhost
* Fixes CSRF for application deployed out of localhost
* Sensible default configurations for deployment environments when `SERVER_HOST` and `FRONTEND_PORT` configurations are provided
* Dynamic configuration on the frontend to specify server deployment location
* example deployment configs for Kubernetes

### CODE REVIEW

File: grai-server/app/the_guide/settings.py [hunk 1]
```diff
@@ -14,21 +15,55 @@
...
+    CORS_ALLOW_ALL_ORIGINS = True
+    CSRF_TRUSTED_ORIGINS = ["http://*", "https://*"]
+else:
+    schemes = ["http", "https"]
+    default_ports = [SERVER_PORT, FRONTEND_PORT]
+    default_hosts = list(set([SERVER_HOST, FRONTEND_HOST]))
+    default_cors_origins = [
+        f"{scheme}://{host}:{port}"
+        for scheme, host, port in product(schemes, default_hosts, default_ports)
+    ]
```
Comment (edlouth) [inline grai-server/app/the_guide/settings.py:53]: Is this crossing all three of these arrays?  I guess it works

File: .gitignore [hunk 2]
```diff
@@ -5,7 +5,12 @@
 .vscode/
 **/.vscode/
 .ipynb_checkpoints
+<<<<<<< Updated upstream
 **/.ipynb_checkpoints/
+=======
+.fleet/
+**/.fleet/
+>>>>>>> Stashed changes
```
Comment (edlouth) [inline .gitignore:13]: Merge error to fix

File: grai-server/app/the_guide/settings.py [hunk 3]
```diff
@@ -88,7 +123,7 @@ def clean_allowed_hosts(val):
     "whitenoise.middleware.WhiteNoiseMiddleware",
     "django.contrib.sessions.middleware.SessionMiddleware",
     "django.middleware.common.CommonMiddleware",
-    "django.middleware.csrf.CsrfViewMiddleware",
+    # "django.middleware.csrf.CsrfViewMiddleware",
```
Comment (edlouth) [inline grai-server/app/the_guide/settings.py:126]: Is this just removing CSRF?

File: grai-server/app/the_guide/settings.py (hunk 1 above)
Comment (ieaves) [reply grai-server/app/the_guide/settings.py:53]: Yes, exactly. It's producing every combination of scheme, host, and port.

File: .gitignore (hunk 2 above)
Comment (ieaves) [reply .gitignore:13]: bah, good catch. thank you.

File: grai-server/app/the_guide/settings.py (hunk 3 above)
Comment (ieaves) [reply grai-server/app/the_guide/settings.py:126]: I'm not sure if it completely disables CSRF but the [docs](https://docs.djangoproject.com/en/4.1/ref/csrf/) describe it providing a specific form of subdomain protection and I wanted to see if that was responsible for the issue I was seeing when deployed to the server.

**Comment by edlouth** [comment]
> Oh fuck yes! That solves the CORS issue on the frontend. I need to figure out why the admin is still generating a CSRF verification issue logging into the admin before we can merge though.
> 
> Screenshot 2022-11-21 at 11 23 53 PM

I think this can be solved by setting `SERVER_HOST=grai.tinglacier.com`

**Comment by ieaves** [comment]
That was annoyingly the configuration I was using in that commit. It's sorted now but the issue ended up being that I only included `scheme://host:port` combinations into `CORS_ALLOWED_ORIGINS` when I needed a record omitting the port suffix. 

There were a few logic bugs I've fixed and cleaned up the CORS_ALLOWED_ORIGINS and CSRF_TRUSTED_ORIGINS so that they aren't maximally permissive.

**Pull request merged**